### PR TITLE
Fix SAML artifact binding failure when tenant is configured as the federated IDP

### PR DESCRIPTION
**Pull request by Wathsara**
## Purpose
Note $subject

### Relates issues
- https://github.com/wso2/product-is/issues/14459

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java [hunk 1]
```diff
@@ -240,4 +249,22 @@ private ArtifactResponse buildArtifactResponse(Response response, ArtifactResolv
...
+     * @throws IdentityException If an error when getting service provider.
+     */
+    private String resolveSpTenantDomain(String tenantDomain, String issuer) throws IdentityException {
+
+        SAMLSSOServiceProviderDO serviceProviderDO = SAMLSSOUtil.getSPConfig(
+                tenantDomain, SAMLSSOUtil.splitAppendedTenantDomain(issuer));
+        if (serviceProviderDO != null) {
+            return serviceProviderDO.getTenantDomain();
+        }
+        return null;
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java:268]: This should break the flow. Doesn't it?

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 2]
```diff
@@ -2690,4 +2691,37 @@ private static SAMLSSOServiceProviderDO getSAMLServiceProviderFromRegistry(Strin
...
+     * @param tenantDomain Service providers tenant domain.
+     * @throws UserStoreException If and error occurs when validating the tenant domain.
+     * @throws IdentityException  If and error occurs when validating the tenant domain.
+     */
+    public static void setSpTenantDomainToThreadLocal(String tenantDomain) throws UserStoreException, IdentityException {
+
+        tenantDomain = validateTenantDomain(tenantDomain);
+        if (tenantDomain != null) {
+            spTenantDomainThreadLocal.set(tenantDomain);
+        }
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java:2717]:
```suggestion
        } else {
            spTenantDomainThreadLocal.remove();
        }
```

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java [hunk 3]
```diff
@@ -108,6 +113,10 @@ public ArtifactResponse resolveArtifact(ArtifactResolve artifactResolve) throws
         }
         catch (Base64DecodingException e) {
             throw new ArtifactBindingException("Error while Base64 decoding SAML2 artifact: " + artifact, e);
+        } catch (UserStoreException e) {
+        throw new ArtifactBindingException("Error while setting the service provider tenant: " + artifact, e);
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java:117]: Format this line.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java [hunk 4]
```diff
@@ -101,12 +104,13 @@ public SignKeyDataHolder(String username) throws IdentityException {
 
             String signWithValue = IdentityUtil.getProperty(
                     SAMLSSOConstants.FileBasedSPConfig.USE_AUTHENTICATED_USER_DOMAIN_CRYPTO);
+            tenantDomain = spTenantDomain;
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java:107]: move this line to 104.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java [hunk 5]
```diff
@@ -101,12 +104,13 @@ public SignKeyDataHolder(String username) throws IdentityException {
...
             if (signWithValue != null && "true".equalsIgnoreCase(signWithValue.trim())) {
                 tenantDomain = userTenantDomain;
+            }
+            tenantID = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+            if (SAMLSSOUtil.getRealmService() != null) {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java:112]: `SAMLSSOUtil.getRealmService() == null` situation should throw a runtime exception as that mean this component is  not started correctly

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java [hunk 6]
```diff
@@ -101,12 +104,13 @@ public SignKeyDataHolder(String username) throws IdentityException {
 
             String signWithValue = IdentityUtil.getProperty(
                     SAMLSSOConstants.FileBasedSPConfig.USE_AUTHENTICATED_USER_DOMAIN_CRYPTO);
+            tenantDomain = spTenantDomain;
             if (signWithValue != null && "true".equalsIgnoreCase(signWithValue.trim())) {
                 tenantDomain = userTenantDomain;
+            }
+            tenantID = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/SignKeyDataHolder.java:111]: This value is always get replaced

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java (hunk 2 above)
Comment (thanujalk) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java:2717]: Just to be in the safe side.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java [hunk 7]
```diff
@@ -240,4 +245,38 @@ private ArtifactResponse buildArtifactResponse(Response response, ArtifactResolv
...
+                String message = "Invalid Tenant Domain : " + tenantDomain;
+                if (log.isDebugEnabled()) {
+                    log.debug(message);
+                }
+                throw IdentityException.error(message);
+            }
+        } catch (UserStoreException e) {
+            String message = "Error occurred while getting tenant ID from tenantDomain " + tenantDomain;
+            log.error(message, e);
+            throw IdentityException.error(message, e);
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java:269]: why are we logging and throwing the exception? If we do like that there is a chance of getting the stack trace printed two times in the console

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java [hunk 8]
```diff
@@ -240,4 +245,37 @@ private ArtifactResponse buildArtifactResponse(Response response, ArtifactResolv
 
         return artifactResponse;
     }
+
+    private void startTenantFlow(String tenantDomain) throws IdentityException {
+
+        if (tenantDomain == null) {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOArtifactResolver.java:251]:
```suggestion
        if (StringUtils.isBlank(tenantDomain)) {
```

**Review by nilasini — APPROVED** [review]

**Review by thanujalk — APPROVED** [review]

**Pull request merged**